### DISCUSSION TITLE
Rule 13?  No UTF8?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ioccc-src/winner#28] [Q&A] [answered]
```
Rule 13

Any C source that fails to compile because of unescaped octets with the high bit set (octet value >= 128) might be rejected. 
```
Is that to say UTF-8 is permitted or some what permitted or `ye take ya chances bub`?

**Comment by xexyl** [comment]
> ```
> Rule 13
> 
> Any C source that fails to compile because of unescaped octets with the high bit set (octet value >= 128) might be rejected. 
> ```
> 
> Is that to say UTF-8 is permitted or some what permitted or `ye take ya chances bub`?

I don't know what the judges have in mind per se but: it certainly means that only if it fails to compile due to that [problem]. So as long as it compiles you're technically not breaking the rule, right?

Something you should keep in mind though ... who's to say that they won't compile it on an older system before this was allowed? :-) So there's that to remember ...

**Reply by SirWumpus** [reply]
Well they're going `-std=gnu17` which allows UTF8 (at least it implies) 5.1.1.2 Translation phases `1. Physical source file multibyte characters are mapped...`

**Comment by lcn2** [comment]
> ```
> 
> Rule 13
> 
> 
> 
> Any C source that fails to compile because of unescaped octets with the high bit set (octet value >= 128) might be rejected. 
> 
> ```
> 
> Is that to say UTF-8 is permitted or some what permitted or `ye take ya chances bub`?

You point out an obvious flaw in the rules that needs to be fixed, @SirWumpus 

> Something you should keep in mind though ... who's to say that they won't compile it on an older system before this was allowed? :-) So there's that to remember ...

Perhaps this too is a flaw, @xexyl, in rule 13.

Should rule 13 be scrapped?

# QUESTION 0

How well would the current winner tools handle UTF-8 in code?  In the process of "scrapping" rule 13, do we need to fix the bin tools or their underlying mkiocccentry repo friends - or are they in good shape to hand things?

You might try as a test, @xexyl, to add some gratuitous UTF-8 character to some entry source and see if `make www` works well.  Don't submit the modified entry as this is just a test.  Just let us know here is all is well or NOT with code that had UTF-8.

# QUESTION 1

If we do "scrap" Rule 13, what fun thing should we put in its place?  Or is there a critical rule we need to add in place of the existing Rule 13?

**Reply by SirWumpus** [reply, marked as answer]
> If we do "scrap" Rule 13, what fun thing should we put in its place? Or is there a critical rule we need to add in place of the existing Rule 13?

Maybe?

* Rule 13 
Rule 13 was scrapped, because it is bad luck to ignore UTF8.

**Comment by lcn2** [comment]
We have opened issue #34 to address Rule 13.

## UPDATE 0

Comments are welcome in issue #34 regarding the proposed fix.

**Comment by lcn2** [comment]
With commit 9dd70f0fc9757045d1825b20cc383fdc3e45a8d8 we will close this discussion with a thank you to @SirWumpus for your input!

**Comment by ilyakurdyukov** [comment]
This rule update does not say how exactly UTF-8 characters will be counted. If long characters are counted as one character in the size calculation, then this can be exploited to significantly expand the program's size. I propose to specify that a UTF-8 character of size N bytes will be counted as N characters when calculating the size.

Also, there should be a restriction that the characters must be defined in the Unicode standard and present in popular fonts (to keep the code readable to the eye). So this rule update is not complete.

**Reply by lcn2** [reply]
> This rule update does not say how exactly UTF-8 characters will be counted. If long characters are counted as one character in the size calculation, then this can be exploited to significantly expand the program's size. I propose to specify that a UTF-8 character of size N bytes will be counted as N characters when calculating the size.

Rule 2 refers to bytes (not characters) only.  A UTF-8 character of size N bytes will count as N bytes.

> Also, there should be a restriction that the characters must be defined in the Unicode standard and present in popular fonts (to keep the code readable to the eye). So this rule update is not complete.

We do not want to force submitters to make their code "readable to the eye".  Sure, the effect might be better ift it was "readable to the eye".  We have made hints in the gudeilines about doing more than presenting a "blob" (search the guidelines for the word "blob").

There will be no "rule" that says that forces the code to "readable to the eye" as we do not want to adversely impact the potential for creativity.
